File: python实现/p1996.py
class Node:
    def __init__(self,val):
        self.val=val
        self.nxt=None
    
def main():
    n,m=map(int,input().split())
    head=Node(1)
    now=head
    for i in range(2,n+1):
        newnode=Node(i)
        now.nxt=newnode
        now=now.nxt
    now.nxt=head
    now=head
    for i in range(n-1):
        for j in range(m-1):
            pre=now
            now=now.nxt
        print(now.val,end=' ')
        pre.nxt=now.nxt
        del now
        now=pre.nxt
    print(now.val)
    
if __name__=='__main__':
    main()

File: p1996.cpp
#include<cstdio>
#include<cstring>

int N,M;
struct Node{
	int val;
	Node* nxt;
};
Node* head;
int main(){
	while(scanf("%d %d",&N,&M)!=EOF){
		head=new Node;
		head->val=1;
		head->nxt=NULL;
		Node* now=head;
		Node* pre=head;
		for(int i=2;i<=N;i++){
			Node* p=new Node;
			p->val=i;
			p->nxt=NULL;
			now->nxt=p;
			pre=now;
			now=p;
		}
		now->nxt=head;
		int count=0;
		while(N--){
			for(int i=1;i<=M;i++){
				pre=now;
				now=now->nxt;
			}
			if(count++) printf(" ");
			printf("%d",now->val);
			pre->nxt=now->nxt;
			delete now;
			now=pre;
		}
		printf("\n");
	}
	return 0;
}
